File: Leetcode 3301-3400/3376/3376.py
# Leetcode 3376: Minimum Time to Break Locks I
# https://leetcode.com/problems/minimum-time-to-break-locks-i/
# Solved on 28th of November, 2025
class Solution:
    def findMinimumTime(self, strength: list[int], k: int) -> int:
        """
        Calculates the minimum time required to break all locks.

        Args:
            strength: A list of integers representing the strength of each lock.
            k: An integer representing the increase in breaking power after breaking each lock.
        Returns:
            An integer representing the minimum total time to break all locks.
        """
        n = len(strength)
        memoStore = {}

        def getMinimum(mask, currentX):
            if mask == (1 << n) - 1:
                return 0

            if mask in memoStore:
                return memoStore[mask]

            minTime = float('inf')

            for i in range(n):
                if not (mask & (1 << i)):
                    timeRequired = (strength[i] + currentX - 1) // currentX
                    remainingTime = getMinimum(mask | (1 << i), currentX + k)
                    totalPathTime = timeRequired + remainingTime

                    if totalPathTime < minTime:
                        minTime = totalPathTime

            memoStore[mask] = minTime
            return minTime

        return getMinimum(0, 1)

File: Leetcode 3301-3400/3376/3376.java
// Leetcode 3376: Minimum Time to Break Locks I
// https://leetcode.com/problems/minimum-time-to-break-locks-i/
// Solved on 28th of November, 2025
import java.util.Arrays;

class Solution {
    /**
     * Calculates the minimum time to break all locks.
     *
     * @param strength A list of integers representing the strength of each lock.
     * @param k An integer representing a factor in the time calculation.
     * @return The minimum time required to break all locks.
     */
    public int findMinimumTime(List<Integer> strength, int k) {
        int n = strength.size();
        int limit = 1 << n;
        int[] dp = new int[limit];
        Arrays.fill(dp, Integer.MAX_VALUE);
        dp[0] = 0;

        for (int mask = 1; mask < limit; mask++) {
            int setBits = Integer.bitCount(mask);
            int x = 1 + (setBits - 1) * k;

            for (int i = 0; i < n; i++) {
                if ((mask & (1 << i)) != 0) {
                    int prevMask = mask ^ (1 << i);
                    int time = (strength.get(i) + x - 1) / x;
                    dp[mask] = Math.min(dp[mask], dp[prevMask] + time);
                }
            }
        }

        return dp[limit - 1];
    }
}